orbotservice/src/main/java/org/torproject/android/service/circumvention/SnowflakeProxyWrapper.kt

```diff
@@ -22,37 +32,59 @@ class SnowflakeProxyWrapper(private val context: Context) {
         if (proxy != null) return
         if (Prefs.limitSnowflakeProxyingWifi() && !hasWifi) return
         if (Prefs.limitSnowflakeProxyingCharging() && !hasPower) return
-        proxy = SnowflakeProxy()
-        val stunServers = BuiltInBridges.getInstance(context)?.snowflake?.firstOrNull()?.ice
-            ?.split(",".toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray() ?: emptyArray()
-        val stunUrl = stunServers[SecureRandom().nextInt(stunServers.size)]
-
-        proxy = SnowflakeProxy()
-        with(proxy!!) {
-            brokerUrl = OrbotService.getCdnFront("snowflake-target-direct")
-            capacity = 1L
-            pollInterval = 120L
-            stunServer = stunUrl
-            relayUrl = OrbotService.getCdnFront("snowflake-relay-url")
-            natProbeUrl = OrbotService.getCdnFront("snowflake-nat-probe")
-            clientConnected = SnowflakeClientConnected { onConnected() }
-            start()
-        }
 
-        if (Prefs.showSnowflakeProxyMessage()) {
-            val message = context.getString(R.string.log_notice_snowflake_proxy_enabled)
-            Handler(context.mainLooper).post {
-                context.applicationContext.showToast(message)
+        CoroutineScope(Dispatchers.IO).launch {
+            val start = Random.nextInt(49152, 65536 - 2)
+
+            for (port in (start..start + 2)) {
+                if (UPnP.openPortUDP(port, OrbotConstants.TAG)) {
+                    mappedPorts.add(port)
+                }
+            }
+
+            if (mappedPorts.size < 3) {
```

Comment:
i might just be missing some context, but if you are only using two ports out of `mappedPorts`, the ephemeral min and max ports, than why does the size need to be less than 3. at a glance, this looks like it should be less than 2, since you're only needing 2 ports here.

Reply:
We need 3 consecutive ports, because that's what Snowflake Proxy wants to function properly when having Capacity set to 1.
In fact, it's Capacity * 2 + 1.
Seems, another comment is necessary. Will add a commit now with all the comments.